clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -352,7 +355,9 @@ private TransactionalRequestResult beginCompletingTransaction(TransactionResult
         enqueueRequest(handler);
 
         // If an epoch bump is required for recovery, initialize the transaction after completing the EndTxn request.
-        if (clientSideEpochBumpRequired) {
+        // If we are upgrading to TV2 transactions on the next transaction, also bump the epoch.
```

Comment:
The order of txn requests is enforced by the queue -- endTxn will happen before InitPid because we queue them here.  The race condition happens is because at this point the endTxn is executing concurrently, it may or may not finish by the time this line runs by this thread, and therefore it may or may not observe the effects of this line.  You can add a sleep before this line and observe the race condition where endTxn finishes before maybeUpdateTransactionV2Enabled is called.

A proper way to do it, is to call this function before enqueuing the request (i.e. move to line 353).  Plus add a comment why it's there an not here.

Reply:
To understand the impact of the race condition:

First, we want to complete the transaction. however, if the end txn completes first then we may additionally start a new transaction here. is this correct?

Reply:
The race condition involved the previous "ready" check that is now removed from the PR. But the idea was that in the race, the EndTxn request would queue and complete before we called the method to maybe upgrade to TV2. In the case that the transaction completes first, we would transition to READY (rather than committing/aborting transaction) and we wouldn't trigger the "upgrade to TV2" (now just `clientSideEpochBumpRequired`) condition. 

Changing the method to not check on ready and check on the parameter eliminated that part of issue, but we also still have the general transition to READY we want to be careful about since we can't easily bump epoch if we are in READY state. 

Right now, since we piggyback on `clientSideEpochBumpRequired` we will always transition to INITIALIZING after completing the transaction since we set that value BEFORE enqueueing the EndTxn request. See https://github.com/apache/kafka/blob/0055ef0a49ca09a350b8b99b0b758d88a9b6555c/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L1244.

TLDR; the issue was not starting a new transaction but ensuring we correctly identify the TV2 upgrade case and bumping the epoch (which could be missed in a race due to idiosyncrasies in the state machine).

Reply:
ah, thanks for the correction, seems like i was thinking about the opposite ordering of events. makes sense